File: repro.py
#!/usr/bin/env python3
"""
Reproducer for a lost TCP reset (RST) on macOS loopback under asyncio.

Summary
-------
On macOS, when a flow-controlled (zero receive window) loopback TCP connection
whose file descriptor is driven by an asyncio ``SelectorEventLoop``
(``KqueueSelector``) is *abortively* closed -- ``SO_LINGER {1, 0}`` then
``close()`` -- the reset is occasionally never delivered to the peer.

The closing side's socket is textbook-correct at ``close()`` time (still
connected, ``SO_LINGER`` set to ``{1, 0}``, unsent bytes in the send buffer),
so per POSIX/BSD semantics ``close()`` must emit a RST. The closing fd is then
released cleanly (it does not linger). Yet the peer's socket stays
``ESTABLISHED`` forever: its ``add_reader`` callback never fires, ``getpeername``
still succeeds, ``SO_ERROR`` is 0, and a ``MSG_PEEK`` ``recv`` returns
``EWOULDBLOCK``. The connection is left half-open with no RST/FIN delivered.

Each iteration reproduces the scenario using only the low-level event-loop file
descriptor APIs (``add_reader`` / ``add_writer`` / ``remove_reader`` /
``remove_writer``) plus raw blocking/non-blocking sockets:

  1. Establish a loopback client + accepted server socket, both non-blocking.
  2. The server never reads, so as the client writes its receive window goes to
     zero and the client's send buffer fills.
  3. The client is registered with the loop (reader + writer) and fills until
     ``send()`` blocks.
  4. The client is unregistered and abortively closed:
     ``setsockopt(SO_LINGER, {1, 0})`` + ``close()``.
  5. The server is registered with ``add_reader``; its callback drains the
     buffered data and then waits for the disconnect (EOF / errno).
  6. If the server's callback never reports the disconnect within TIMEOUT, the
     reset was not delivered -- the bug. Diagnostics proving the socket is
     still ESTABLISHED are printed.

Observed
--------
- macOS 14 (Kqueue): reproduces, rarely (order 1 in 10^5-10^6 iterations).
- macOS 15 (Kqueue): not observed in ~10^6 iterations (much rarer, if present).
- Linux (epoll): never observed.

Usage
-----
    python3 repro.py [SECONDS]      # default 300s

Exit status is non-zero if any undelivered (half-open) close was observed.
"""
import asyncio
import errno
import platform
import selectors
import socket
import struct
import sys
import time

TIMEOUT = 1.0
LINGER_ZERO = struct.pack("ii", 1, 0)


def make_pair():
    """A connected loopback (client, server) socket pair, both non-blocking."""
    lsock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    lsock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    lsock.bind(("127.0.0.1", 0))
    lsock.listen(1)
    client = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    client.connect(lsock.getsockname())
    server, _ = lsock.accept()
    lsock.close()
    client.setblocking(False)
    server.setblocking(False)
    return client, server


def describe(sock):
    """Prove whether the socket still looks like a live, idle connection."""
    try:
        peer = f"getpeername={sock.getpeername()}"
    except OSError as e:
        peer = f"getpeername->{errno.errorcode.get(e.args[0], e.args[0])}"
    try:
        so = sock.getsockopt(socket.SOL_SOCKET, socket.SO_ERROR)
    except OSError as e:
        so = f"err{e.args[0]}"
    try:
        d = sock.recv(1, socket.MSG_PEEK)
        peek = "EOF" if d == b"" else f"{len(d)}bytes"
    except OSError as e:
        peek = errno.errorcode.get(e.args[0], str(e.args[0]))
    return f"{peer} SO_ERROR={so} MSG_PEEK={peek}"


async def one_iteration(loop):
    client, server = make_pair()
    cfd, sfd = client.fileno(), server.fileno()

    # Client is a reader (as a normal connection is) and a writer while it has
    # data to send; fill until the send buffer blocks against the non-reading
    # server (receive window -> zero).
    loop.add_reader(cfd, lambda: None)
    fill_done = loop.create_future()
    chunk = b"x" * 65536

    def do_write():
        try:
            while True:
                client.send(chunk)
        except OSError:
            loop.remove_writer(cfd)
            if not fill_done.done():
                fill_done.set_result(None)

    loop.add_writer(cfd, do_write)
    await fill_done

    # Abortively close the client: unregister it, then reset.
    loop.remove_reader(cfd)
    loop.remove_writer(cfd)
    client.setsockopt(socket.SOL_SOCKET, socket.SO_LINGER, LINGER_ZERO)
    client.close()

    # The server starts reading; its callback drains buffered data and then
    # waits for the disconnect.
    got = loop.create_future()

    def on_server_readable():
        try:
            data = server.recv(1 << 20)
        except OSError as e:
            if not got.done():
                got.set_result(errno.errorcode.get(e.args[0], str(e.args[0])))
            return
        if data == b"":
            if not got.done():
                got.set_result("EOF")
        # else: drained buffered data; keep waiting for the disconnect

    loop.add_reader(sfd, on_server_readable)
    try:
        await asyncio.wait_for(got, TIMEOUT)
        outcome = "delivered"
    except asyncio.TimeoutError:
        outcome = "UNDELIVERED: " + describe(server)
    finally:
        try:
            loop.remove_reader(sfd)
        except Exception:
            pass
        server.close()
        try:
            client.close()
        except Exception:
            pass
    return outcome


async def main_coro(run_seconds):
    loop = asyncio.get_running_loop()
    delivered = 0
    undelivered = 0
    deadline = time.monotonic() + run_seconds
    while time.monotonic() < deadline:
        res = await one_iteration(loop)
        if res == "delivered":
            delivered += 1
        else:
            undelivered += 1
            print(f"  [{delivered + undelivered}] {res}", flush=True)
    return delivered, undelivered


def main():
    run_seconds = float(sys.argv[1]) if len(sys.argv) > 1 else 300.0
    print(f"platform: {platform.platform()}  python: {sys.version.split()[0]}")
    print(f"selector: {type(selectors.DefaultSelector()).__name__}")
    print(f"run={run_seconds}s timeout={TIMEOUT}s\n")

    delivered, undelivered = asyncio.run(main_coro(run_seconds))
    total = delivered + undelivered
    print(f"\ndelivered   = {delivered}/{total}")
    print(f"UNDELIVERED = {undelivered}/{total}")
    sys.exit(1 if undelivered else 0)


if __name__ == "__main__":
    main()

File: c/repro.c
/*
 * Pure-C reproducer for the lost TCP reset (RST) on macOS loopback, using the
 * raw kqueue/kevent and socket syscalls only (no libraries beyond libc).
 *
 * When a flow-controlled (zero receive window) loopback TCP connection is
 * abortively closed -- setsockopt(SO_LINGER {1,0}) then close() -- the reset is
 * occasionally never delivered to the peer. The peer's socket stays ESTABLISHED
 * forever: its EVFILT_READ never fires, getpeername() still succeeds,
 * SO_ERROR == 0, and a MSG_PEEK recv() returns EWOULDBLOCK.
 *
 * Mirrors the "deferred close" pattern that reproduces most often: a kevent
 * poll cycle runs between deleting the client fd's filters and abortively
 * closing it.
 *
 * Each iteration:
 *   1. Establish a loopback client + accepted server socket, both non-blocking.
 *   2. Register the client (EVFILT_READ + EVFILT_WRITE); fill via write-
 *      readiness until send() blocks against the non-reading server.
 *   3. EV_DELETE the client's filters, run one kevent poll cycle, then
 *      abortively close it.
 *   4. Register the server (EVFILT_READ); drain the buffered data, then wait
 *      for the disconnect (EOF / errno).
 *   5. If the disconnect never arrives within TIMEOUT, the reset was lost.
 *
 * Build:  cc -O2 -o repro repro.c
 * Run:    ./repro [SECONDS] [stop]
 *           SECONDS: how long to run (default 300)
 *           stop:    if present, exit at the first lost RST (for packet capture)
 *
 * Exit status is non-zero if any lost RST is observed.
 */
#include <arpa/inet.h>
#include <errno.h>
#include <fcntl.h>
#include <netinet/in.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/event.h>
#include <sys/socket.h>
#include <sys/time.h>
#include <time.h>
#include <unistd.h>

#define TIMEOUT_SEC 1.0

static double now_sec(void) {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return (double)ts.tv_sec + (double)ts.tv_nsec / 1e9;
}

static void set_nonblocking(int fd) {
    int fl = fcntl(fd, F_GETFL, 0);
    fcntl(fd, F_SETFL, fl | O_NONBLOCK);
}

static void make_pair(int *client_out, int *server_out) {
    int lsock = socket(AF_INET, SOCK_STREAM, 0);
    int one = 1;
    setsockopt(lsock, SOL_SOCKET, SO_REUSEADDR, &one, sizeof(one));
    struct sockaddr_in a;
    memset(&a, 0, sizeof(a));
    a.sin_family = AF_INET;
    a.sin_port = 0;
    a.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
    bind(lsock, (struct sockaddr *)&a, sizeof(a));
    listen(lsock, 1);

    struct sockaddr_in bound;
    socklen_t blen = sizeof(bound);
    getsockname(lsock, (struct sockaddr *)&bound, &blen);

    int client = socket(AF_INET, SOCK_STREAM, 0);
    connect(client, (struct sockaddr *)&bound, sizeof(bound));
    int server = accept(lsock, NULL, NULL);
    close(lsock);

    set_nonblocking(client);
    set_nonblocking(server);
    *client_out = client;
    *server_out = server;
}

static void kev_change(int kq, int fd, int16_t filter, uint16_t flags) {
    struct kevent ev;
    EV_SET(&ev, fd, filter, flags, 0, 0, NULL);
    kevent(kq, &ev, 1, NULL, 0, NULL);
}

static int kev_poll(int kq, struct kevent *evs, int max, double timeout) {
    struct timespec ts;
    ts.tv_sec = (time_t)timeout;
    ts.tv_nsec = (long)((timeout - (double)(time_t)timeout) * 1e9);
    int n = kevent(kq, NULL, 0, evs, max, &ts);
    return n < 0 ? 0 : n;
}

static int local_port(int fd) {
    struct sockaddr_in a;
    socklen_t l = sizeof(a);
    if (getsockname(fd, (struct sockaddr *)&a, &l) == 0)
        return ntohs(a.sin_port);
    return -1;
}

static int peer_port(int fd) {
    struct sockaddr_in a;
    socklen_t l = sizeof(a);
    if (getpeername(fd, (struct sockaddr *)&a, &l) == 0)
        return ntohs(a.sin_port);
    return -1;
}

/* returns 1 if the abort was NOT delivered (the hang), 0 if delivered. */
static int one_iteration(int kq, char *buf, size_t buflen, char *desc,
                         size_t desclen) {
    int client, server;
    make_pair(&client, &server);
    static char chunk[65536];
    struct kevent evs[16];

    /* Register the client for read+write; fill until send() blocks. */
    kev_change(kq, client, EVFILT_READ, EV_ADD);
    kev_change(kq, client, EVFILT_WRITE, EV_ADD);
    int filled = 0;
    double fill_deadline = now_sec() + TIMEOUT_SEC;
    while (!filled && now_sec() < fill_deadline) {
        int n = kev_poll(kq, evs, 16, 0.5);
        for (int i = 0; i < n; i++) {
            if ((int)evs[i].ident == client && evs[i].filter == EVFILT_WRITE) {
                while (send(client, chunk, sizeof(chunk), 0) > 0) {
                }
                kev_change(kq, client, EVFILT_WRITE, EV_DELETE);
                filled = 1;
            }
        }
    }
    if (!filled) {
        kev_change(kq, client, EVFILT_READ, EV_DELETE);
        close(client);
        close(server);
        return 0;
    }

    /* Deferred abort: delete the client's remaining filter, run one poll cycle,
     * then abortively close it. */
    kev_change(kq, client, EVFILT_READ, EV_DELETE);
    kev_poll(kq, evs, 16, 0.0); /* the poll cycle between unregister and close */
    struct linger lg = {1, 0};
    setsockopt(client, SOL_SOCKET, SO_LINGER, &lg, sizeof(lg));
    int lport = local_port(server);
    int pport = peer_port(server);
    close(client);

    /* Register the server; drain buffered data, then wait for the disconnect. */
    kev_change(kq, server, EVFILT_READ, EV_ADD);
    int hung = 1;
    double deadline = now_sec() + TIMEOUT_SEC;
    for (;;) {
        double rem = deadline - now_sec();
        if (rem <= 0) {
            hung = 1;
            break;
        }
        int n = kev_poll(kq, evs, 16, rem);
        int got = 0;
        for (int i = 0; i < n; i++) {
            if ((int)evs[i].ident == server && evs[i].filter == EVFILT_READ) {
                ssize_t r = recv(server, buf, buflen, 0);
                if (r <= 0) { /* ECONNRESET / EOF */
                    hung = 0;
                    got = 1;
                    break;
                }
                /* else: drained buffered data; keep waiting */
            }
        }
        if (got)
            break;
    }

    if (hung) {
        int so = 0;
        socklen_t sl = sizeof(so);
        getsockopt(server, SOL_SOCKET, SO_ERROR, &so, &sl);
        char pk[32];
        ssize_t r = recv(server, buf, 1, MSG_PEEK);
        if (r < 0)
            snprintf(pk, sizeof(pk), "errno %d", errno);
        else if (r == 0)
            snprintf(pk, sizeof(pk), "EOF");
        else
            snprintf(pk, sizeof(pk), "%zdbytes", r);
        snprintf(desc, desclen, "%d<->%d SO_ERROR=%d MSG_PEEK=%s", lport, pport,
                 so, pk);
    }
    close(server);
    return hung;
}

int main(int argc, char **argv) {
    double run_seconds = argc > 1 ? atof(argv[1]) : 300.0;
    int stop_on_hang = argc > 2;

    printf("c repro: raw kqueue/kevent + sockets (no libraries)\n");
    printf("run=%.0fs timeout=%.0fs (deferred close)%s\n\n", run_seconds,
           TIMEOUT_SEC, stop_on_hang ? " [stop at first hang]" : "");
    fflush(stdout);

    int kq = kqueue();
    size_t buflen = 1 << 20;
    char *buf = malloc(buflen);
    unsigned long long delivered = 0, undelivered = 0;
    char desc[128];
    double start = now_sec();

    while (now_sec() - start < run_seconds) {
        if (one_iteration(kq, buf, buflen, desc, sizeof(desc))) {
            undelivered++;
            printf("  [%llu] UNDELIVERED: %s\n", delivered + undelivered, desc);
            fflush(stdout);
            if (stop_on_hang)
                break;
        } else {
            delivered++;
        }
    }
    close(kq);

    unsigned long long total = delivered + undelivered;
    printf("\ndelivered   = %llu/%llu\n", delivered, total);
    printf("UNDELIVERED = %llu/%llu\n", undelivered, total);
    return undelivered > 0 ? 1 : 0;
}
